odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+    >>> x = make_avsc_object(test_dshape, name="my_record", namespace="com.blaze.odo")
+    >>> x == {'fields': [{'name': 'letter', 'type': 'string'},
+    ...                {'name': 'value', 'type': ['null', 'int']}],
+    ...                 'name': 'my_record',
+    ...                 'namespace': 'com.blaze.odo',
+    ...                  'type': 'record'}
+    True
+    >>> test_dshape = '''
+    ...    var * {
+    ...      field_1: int32,
+    ...      field_2: string,
+    ...      field_3: ?int64,
+    ...      features: map[string, float64],
+    ...      words: var * string,
+    ...      nested_record: var * {field_1: int64, field_2: float32}
+    ...    }
+    ... '''
+    >>> x = make_avsc_object(test_dshape, name="my_record", namespace="com.blaze.odo")
+    >>> x == {'fields': [{'name': 'field_1', 'type': 'int'},
+    ...            {'name': 'field_2', 'type': 'string'},
+    ...            {'name': 'field_3', 'type': ['null', 'long']},
+    ...            {'name': 'features', 'type': {'type': 'map', 'values': 'double'}},
+    ...            {'name': 'words', 'type': {'items': 'string', 'type': 'array'}},
+    ...            {'name': 'nested_record',
+    ...             'type': {'items': {'fields': [{'name': 'field_1',
+    ...                                            'type': 'long'},
+    ...                                           {'name': 'field_2',
+    ...                                            'type': 'float'}],
+    ...                                'name': 'my_recordd0d1',
+    ...                                'namespace': 'com.blaze.odo',
+    ...                                'type': 'record'},
+    ...                      'type': 'array'}}],
+    ... 'name': 'my_record',
+    ... 'namespace': 'com.blaze.odo',
+    ... 'type': 'record'}
+    True
+    """
+
+    try:
+        assert depth >= 0
+    except AssertionError:
+        raise ValueError("depth argument must be >= 0")
+
+    #parse string to datashape object if necessary
+    if isinstance(ds, six.string_types) or isinstance(ds, six.text_type):
+        ds = dshape(ds)
+    if isinstance(ds, ct.DataShape):
+        if depth>0:
+            assert isinstance(ds.parameters[0], ct.Var), "Cannot support fixed length substructure in Avro schemas"
+            return {"type": "array", "items": make_avsc_object(ds.measure, name=name+"d%d" % depth, namespace=namespace, depth=depth+1)}
+        elif depth==0:
+            ds = ds.measure
+
+    if isinstance(ds, ct.Record):
+        return {
+            "type": "record",
+            "namespace": namespace,
+            "name": name,
+            "fields": [{"type": make_avsc_object(typ, name=name+"d%d" % depth, namespace=namespace, depth=depth+1),
+                        "name": n} for (typ, n) in zip(ds.measure.types, ds.measure.names)]
+
+        }
+    if isinstance(ds, ct.Map):
+        assert ds.key == ct.string, "Avro map types only support string keys.  Cannot form map with key type %s" % ds.key
```

Comment:
This should be converted raising a `ValueError`, since `ds` is user input.  `assert` statements may not be run -- for instance, when Python generates "optimized" bytecode.